### DISCUSSION TITLE
SignIn callback error when using NextAuth with [OAUTH_CALLBACK_ERROR]

### DISCUSSION THREAD

**Discussion by HosseinEnsafi** [nextauthjs/next-auth#3919] [Help]
### Question 💬

I'm trying to handle authentication with gmail using NextAuth and I'm new to it.
when i press the singin button i face **expected 200 OK, got: 403 Forbidden** message

and this code is came from config file for nextAuth:

```
import NextAuth from "next-auth";
import GoogleProvider from "next-auth/providers/google";

export default NextAuth({
  // Configure one or more authentication providers
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    }),
    // ...add more providers here
  ],
  callbacks: {
    async session({ session, token }) {
      session.user.tag = session.user.name
        .split(" ")
        .join("")
        .toLocaleLowerCase();

      session.user.uid = token.sub;
      return session;
    },
  },
  secret: process.env.NEXTAUTH_SECRET,
});
```

and this is full error message:
[next-auth][error][OAUTH_CALLBACK_ERROR] 
https://next-auth.js.org/errors#oauth_callback_error expected 200 OK, got: 403 Forbidden {
  error: {
    message: 'expected 200 OK, got: 403 Forbidden',
    stack: 'OPError: expected 200 OK, got: 403 Forbidden\n' +
      '    at processResponse (G:\\vs Code\\React\\Clones\\ilw-clones\\twitter-2\\node_modules\\openid-client\\lib\\helpers\\process_response.js:41:11)\n' +
      '    at G:\\vs Code\\React\\Clones\\ilw-clones\\twitter-2\\node_modules\\openid-client\\lib\\helpers\\issuer.js:47:22\n' +
      '    at runMicrotasks (<anonymous>)\n' +
      '    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n' +
      '    at async Issuer.queryKeyStore (G:\\vs Code\\React\\Clones\\ilw-clones\\twitter-2\\node_modules\\openid-client\\lib\\helpers\\issuer.js:83:20)\n' +
      '    at async Client.validateJWT (G:\\vs Code\\React\\Clones\\ilw-clones\\twitter-2\\node_modules\\openid-client\\lib\\client.js:1048:14)\n' +
      '    at async Client.validateIdToken (G:\\vs Code\\React\\Clones\\ilw-clones\\twitter-2\\node_modules\\openid-client\\lib\\client.js:765:49)\n' +
\openid-client\lib\helpers\issuer.js:47:22
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
  code: undefined
}
Error: Requested and resolved page mismatch: /../public/twitter-icon-svg-28.jpg /public/twitter-icon-svg-28.jpg
    at Object.normalizePagePath (G:\vs Code\React\Clones\ilw-clones\twitter-2\node_modules\next\dist\server\normalize-page-path.js:35:15)
    at DevServer.hasPage (G:\vs Code\React\Clones\ilw-clones\twitter-2\node_modules\next\dist\server\dev\next-dev-server.js:331:54)
    at DevServer._beforeCatchAllRender (G:\vs Code\React\Clones\ilw-clones\twitter-2\node_modules\next\dist\server\dev\next-dev-server.js:355:28)
    at async Object.fn (G:\vs Code\React\Clones\ilw-clones\twitter-2\node_modules\next\dist\server\dev\next-dev-server.js:602:21)
    at async Router.execute (G:\vs Code\React\Clones\ilw-clones\twitter-2\node_modules\next\dist\server\router.js:228:32)       
    at async DevServer.run (G:\vs Code\React\Clones\ilw-clones\twitter-2\node_modules\next\dist\server\base-server.js:598:29)   
    at async DevServer.run (G:\vs Code\React\Clones\ilw-clones\twitter-2\node_modules\next\dist\server\dev\next-dev-server.js:444:20)
    at async DevServer.handleRequest (G:\vs Code\React\Clones\ilw-clones\twitter-2\node_modules\next\dist\server\base-server.js:305:20)
    at async imageOptimizer (G:\vs Code\React\Clones\ilw-clones\twitter-2\node_modules\next\dist\server\image-optimizer.js:290:17)
    at async Router.execute (G:\vs Code\React\Clones\ilw-clones\twitter-2\node_modules\next\dist\server\router.js:228:32)       

**I Know there must be an issue with openid-client package but I cant still solve the problem**

**Comment by nishanthrk** [comment]
I think openid-client has some issue with node 14 vercel by default runs it on 14, I'm not facing this issue in node 16

**Reply by HosseinEnsafi** [reply]
thank you very much ,
it was sanction issue i think , because Iranian people are forbidden to use Google Could .
problem was solved by using github services

**Reply by mostafabazyar** [reply]
haji pashmam man 2 rooze daram migam chera nemitonam login konam pas dalilesh ine fk mersi hossein jan

**Reply by mohammadkhakshoor** [reply]
Hi there
i faced with the same issue 
but the point is , i could solve this issue and have a work around by using the verson beta(V5) of the next-auth which now is being called [auth.js](https://authjs.dev/getting-started/installation)
and i got the same error using the next auth v4.*.* 
not really sure why it works with v5 but not v4 but i the mecanism of v5 is a bit diffrent with v4
hope you could solve this after almost 2 years 
cheers

**Reply by surfapps** [reply]
> Hi there i faced with the same issue but the point is , i could solve this issue and have a work around by using the verson beta(V5) of the next-auth which now is being called [auth.js](https://authjs.dev/getting-started/installation) and i got the same error using the next auth v4._._ not really sure why it works with v5 but not v4 but i the mecanism of v5 is a bit diffrent with v4 hope you could solve this after almost 2 years cheers

Thank you very much. You saved my day. The problem is solved in beta(V5).
It is worth noting for all of the persons in this discussion to know that there is no sanction for Google sign-in.

**Reply by mohammadkhakshoor** [reply]
> haji pashmam man 2 rooze daram migam chera nemitonam login konam pas dalilesh ine fk mersi hossein jan

Hi,
If it was about sanctions, it shouldn’t work with V5 of auth.js.
It works fine with V5, but it does not work with V4.

**Comment by HosseinEnsafi** [comment]
Man haft rooz vali akhar drst nashod
Ba GitHub anjam bde

دریافت Outlook برای Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by mostafabazyar** [comment]
Man aval ba GitHub anjam dadam bara hamin nafahmidam moshkel az ine akharesh bayad mail server run konam barash fk konam Sherkat gofte google hatman bashe

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by imrankhanjoya** [comment]
Getting this error with twitter login button please help:
[next-auth][error][OAUTH_CALLBACK_ERROR] 
https://next-auth.js.org/errors#oauth_callback_error expected 200 OK, got: 403 Forbidden {
  error: OPError: expected 200 OK, got: 403 Forbidden
      at processResponse (/private/var/www/my_nextjs/lynkuidadmin/node_modules/openid-client/lib/helpers/process_response.js:41:11)
      at Client.userinfo (/private/var/www/my_nextjs/lynkuidadmin/node_modules/openid-client/lib/client.js:1235:18)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async oAuthCallback (/private/var/www/my_nextjs/lynkuidadmin/node_modules/next-auth/core/lib/oauth/callback.js:145:17)
      at async Object.callback (/private/var/www/my_nextjs/lynkuidadmin/node_modules/next-auth/core/routes/callback.js:50:11)
      at async NextAuthHandler (/private/var/www/my_nextjs/lynkuidadmin/node_modules/next-auth/core/index.js:201:28)
      at async NextAuthNextHandler (/private/var/www/my_nextjs/lynkuidadmin/node_modules/next-auth/next/index.js:23:19)
      at async /private/var/www/my_nextjs/lynkuidadmin/node_modules/next-auth/next/index.js:59:32
      at async Object.apiResolver (/private/var/www/my_nextjs/lynkuidadmin/node_modules/next/dist/server/api-utils/node.js:366:9)
      at async DevServer.runApi (/private/var/www/my_nextjs/lynkuidadmin/node_modules/next/dist/server/next-server.js:481:9) {
    name: 'OAuthCallbackError',
    code: undefined
  },
  providerId: 'twitter',
  message: 'expected 200 OK, got: 403 Forbidden'
}

**Comment by HosseinEnsafi** [comment]
سلام شرمنده پیامت نگرفتم اون روز نمیدونم دیگه چطور باید حل بشه خودم هم توش موندم اون زمان البته برای تمرین من زده بودم که نشد

**Comment by yahya-aghdam** [comment]
کلا این کتابخونه مشکل داره
من دستی یه کتابخونه دگ نوشتم :|

**Reply by mostafabazyar** [reply]
منم همین کارو کردم با نود زدم یکی دیگه

**Reply by iliaghz** [reply]
دوستان میشه لطفا کتابخونتون رو به اشتراک بزارید ممنون میشم 🙏

**Reply by Christosanesti** [reply]
Supabase

**Comment by amintabrizi** [comment]
دوستان برای رفع مشکل راهکاری پیدا نکردید؟

**Comment by Kasra-Saeid** [comment]
Next auth using google provider does not work for Iranian developers, we  are facing many difficulties. can you solve the issue? all above developers are from Iran and some of them had to not use Next auth as authentication method to use google as provider.

**Reply by Christosanesti** [reply]
This can't be the case. The package runs locally and doesn't need IP verification. The issue (most likely) stems from webpack/react/next environment. 

try the debug option in the auth.ts file

**Comment by Kasra-Saeid** [comment]
I think the category should be issue not help!

**Comment by HosseinEnsafi** [comment]
I think , found the solution . there is a `rewrite` property in nextjs.config.js that could be used in nextAuth and all other auth providers. this specially works for Iranian people who are under google sanctions . use your system proxy  in your  

```
/** @type {import('next').NextConfig} */
// next.config.js
const nextConfig = {
  async rewrites() {
    return [
      {
        source: "/:path*",
        destination: "http://127.0.0.1:8000/:path*", // Adjust if you have a different port or service
      },
    ]
  },
}

export default nextConfig

```

this will automatically apply to all requests, including your development server. hope you will not get that damn 403 error status anymore .

**Reply by Christosanesti** [reply]
use 
debug: true 

and find why are you getting a 403. NextAuth is way too abstract and finding the glitch is a hassle.

**Reply by HosseinEnsafi** [reply]
the question had been resolved long time ago
one word: sanctions

**Comment by surfapps** [comment]
You can use auth.js beta(V5) which is the new version of next-auth.
I used the instruction like the following tutorial link and everything is OK
https://authjs.dev/getting-started/installation
